Error Specificity in DNN Module

shubhanshu02 opened the pull request:
This patchset introduces error codes from `libavutil\error.h` in the Deep Neural Networks module to enhance the error handling.

### Usage

- `DNN_GENERIC_ERROR` for backend framework-specific errors
- `AVERROR(ENOSYS)` for missing features

shubhanshu02 committed:
lavfi/dnn: Error Specificity in Native Backend Layers

This commit returns specific error codes from the execution
functions in the Native Backend layers instead of DNN_ERROR.

Signed-off-by: Shubhanshu Saxena <shubhanshu.e01@gmail.com>

shubhanshu02 committed:
lavfi/dnn_io_proc: Return Specific Error Codes

This commit returns specific error codes from the functions in the
dnn_io_proc instead of DNN_ERROR.

Signed-off-by: Shubhanshu Saxena <shubhanshu.e01@gmail.com>

shubhanshu02 committed:
lavfi/dnn_backend_openvino: Return Specific Error Codes

Switch to returning specific error codes or DNN_GENERIC_ERROR
when an error is encountered. For OpenVINO API errors, currently
DNN_GENERIC_ERROR is returned.

shubhanshu02 committed:
lavfi/dnn_backend_tf: Return Specific Error Codes

Switch to returning specific error codes or DNN_GENERIC_ERROR
when an error is encountered. For TensorFlow C API errors, currently
DNN_GENERIC_ERROR is returned.

Signed-off-by: Shubhanshu Saxena <shubhanshu.e01@gmail.com>

shubhanshu02 committed:
lavfi/dnn_backend_native: Return Specific Error Codes

Switch to returning specific error codes or DNN_GENERIC_ERROR
when an error is encountered.

Signed-off-by: Shubhanshu Saxena <shubhanshu.e01@gmail.com>

shubhanshu02 committed:
lavfi/dnn_backend_common: Return specific error codes

Switch to returning specific error codes or DNN_GENERIC_ERROR
when an error is encountered in the common DNN backend functions.

Signed-off-by: Shubhanshu Saxena <shubhanshu.e01@gmail.com>

shubhanshu02 committed:
libavfilter: Prepare to handle specific error codes in DNN Filters

This commit prepares the filter side to handle specific error codes
from the DNN backends instead of current DNN_ERROR.

Signed-off-by: Shubhanshu Saxena <shubhanshu.e01@gmail.com>

guoyejun commented on libavfilter/dnn_filter_common.c:
```diff
@@ -106,18 +106,18 @@ int ff_dnn_set_classify_post_proc(DnnContext *ctx, ClassifyPostProc post_proc)
 }
 
-DNNReturnType ff_dnn_get_input(DnnContext *ctx, DNNData *input)
+int ff_dnn_get_input(DnnContext *ctx, DNNData *input)
```
will there be build warning since ctx->model->get_input still returns DNNReturnType?

anyway, i think this is a good idea to decouple the change.

shubhanshu02 replied:
I didn't see any build warning related to `libavfilter` after this commit (no warning earlier also).

Decouple the change means?

guoyejun replied:
i mean to separate the changes into small commits.

shubhanshu02 replied:
Oh, got it! Currently, I have migrated one module per commit which is a bigger change. Migrating one function each commit instead will lead to smaller commits.

guoyejun replied:
it's ok for the current patch set

shubhanshu02 replied:
Oh, okay. :smile:

guoyejun commented on libavfilter/dnn_interface.h:
```diff
@@ -30,6 +30,8 @@
 
+#define DNN_GENERIC_ERROR FFERRTAG('D','N','N','!')
+
 typedef enum {DNN_SUCCESS, DNN_ERROR} DNNReturnType;
```
how about to remove DNNReturnType?

returns 0 for success, and negative values for errors.

shubhanshu02 replied:
Sure. That's seems better since `DNN_ERROR` isn't used anywhere after these changes.

shubhanshu02 committed:
libavfilter: Remove DNNReturnType from DNN Module

This patch removes all occurences of DNNReturnType from the DNN module.
This commit replaces DNN_SUCCESS by 0 (essentially the same), so the
functions with DNNReturnType now return 0 in case of success, the negative
values otherwise.

Signed-off-by: Shubhanshu Saxena <shubhanshu.e01@gmail.com>

Semmer2 commented:
LGTM, if @guoyejun has no more comments, its' OK to send them out. :)

shubhanshu02 commented:
Thanks for reviewing. Can you also check the tests' logs that are failing - `ubuntu/focal` and `ubuntu/bionic`?

Semmer2 commented:
> Thanks for reviewing. Can you also check the tests' logs that are failing - `ubuntu/focal` and `ubuntu/bionic`?

We believe it's a CI system issue, not from your code. It's OK to send them to maillist.

shubhanshu02 commented:
Sure, I'll send the patchset to the mailing list. Thanks.

shubhanshu02 commented:
Thanks for reviewing.
